dvc/repo/experiments/queue/celery.py

```diff
@@ -322,6 +326,19 @@ def follow(
         for line in self.proc.follow(entry.stash_rev, encoding):
             ui.write(line, end="")
 
+        MAX_RETRY = 5
+        for _ in range(MAX_RETRY):
+            for _, queue_entry in self._iter_done_tasks():
+                if queue_entry == entry:
+                    logger.debug("entry %s finished", entry.stash_rev)
+                    return
+            time.sleep(1)
+        logger.warning(
+            "Post process experiment %s time out with max retries %d.",
+            entry.stash_rev,
+            MAX_RETRY,
+        )
```

Comment:
This doesn't belong in `follow()`, this will break the use case where we the user is using `queue logs -f` and ctrl-c's to stop viewing the logs (it should exit without waiting for the underlying task to finish).

If there are places that use `follow()` but we need to actually wait for the entire task to finish, we should really be doing something like
```
celery_queue.follow(entry)
celery_queue.get_result(entry)
```

`get_result()` has better logic for waiting until the given entry is completed, we should avoid this kind of busy-wait `sleep()` whenever possible

Reply:
Yes, the problem is that the `get_result` is leaky, we might get the result before the tasks are complete. 

https://github.com/iterative/dvc/blob/c01583f3dd0084f6386332982f8a1eec32db762c/dvc/repo/experiments/queue/celery.py#L255-L265

Here we directly look into the result directly without checking the `AysncResult.ready()` or waiting until``AysncResult.get()` is returned. And this is where the problem is.